File: src/main/java/net/frozenorb/potpvp/nametag/provider/NameTagProvider.java
package net.frozenorb.potpvp.nametag.provider;

import lombok.AllArgsConstructor;
import lombok.Getter;
import net.frozenorb.potpvp.PotPvPRP;
import org.bukkit.entity.Player;
import net.frozenorb.potpvp.nametag.construct.NameTagInfo;
import xyz.refinedev.spigot.utils.CC;

@Getter
@AllArgsConstructor
public abstract class NameTagProvider {

    private final PotPvPRP plugin = PotPvPRP.getInstance();

    private final String name;
    private final int weight;

    public abstract NameTagInfo fetchNameTag(Player toRefresh, Player refreshFor);

    public NameTagInfo createNameTag(String prefix, String suffix) {
        return (plugin.getNameTagEngine().getOrCreate(CC.translate(prefix), CC.translate(suffix)));
    }
}

File: src/main/kotlin/net/frozenorb/potpvp/kt/nametag/NametagProvider.kt
package net.frozenorb.potpvp.kt.nametag

import net.frozenorb.potpvp.PotPvPRP
import org.bukkit.entity.Player

abstract class NametagProvider

constructor(val name: String, val weight: Int) {

    abstract fun fetchNametag(toRefresh: Player, refreshFor: Player): NametagInfo

    class DefaultNametagProvider : NametagProvider("Default Provider", 0) {
        override fun fetchNametag(toRefresh: Player, refreshFor: Player): NametagInfo {
            return createNametag("", "")
        }
    }

    companion object {
        @JvmStatic
        fun createNametag(prefix: String, suffix: String): NametagInfo {
            return PotPvPRP.getInstance().nameTagHandler.getOrCreate(prefix, suffix)
        }
    }

}